src/clj/puppetlabs/puppetserver/certificate_authority.clj

```diff
@@ -475,6 +486,43 @@
    (utils/subject-key-identifier
      ca-public-key false)])
 
+(schema/defn read-compile-masters
+    "Returns a list of compile masters from the specified file.
+     Note that this method is agnostic of the file contents which could either
+     be node names (FQDNs) or certificate serial numbers so long as they are
+     organized as one item per line.
+     Likewise this method could/will be re-used"
+    [cm-file :- schema/Str]
+    (line-seq (io/reader cm-file)))
+
+(schema/defn generate-compile-master-serials
+  "Given a file containing names of all compile master nodes
+   it will create a file containing serial numbers of their certificates.
+   Both files will have names/serial numbers on separate lines.
+   In case the compile masters file does not exist it will create an 
+   empty file of compile master serial numbers."
+   [compile-masters-file :- schema/Str
+    compile-masters-serials-file :- schema/Str
```

Comment:
What do we want to do when the serials-file already exists? Do we want to overwrite it? Do we only want to load the certs that aren't already in it (may cause issues with the `allow-duplicate-certnames` setting)? Do we want to load all the certs but only add serials that do not already exist in it? What happens if we can't find a cert for an infra node?

Reply:
Yes the serials-file would be regenerated (overwritten) every time compile masters change as it is supposed to maintain the 'current set of compile masters'. Allow-duplicate-certnames refers to whether CA will issue multiple certificates with same subject names so hopefully should not matter. And I hope/assume current interfaces (UI or cli) do not allow for revoking same certificate more than once. Else this could be an issue for existing CRLs as well depending on how CA deals with same serial getting revoked more than once. 
The case of not being able to find a certificate for an infra node would be an issue. I think there is a backup option of figuring that assuming inventory.txt exists and is reliable.

Reply:
> And I hope/assume current interfaces (UI or cli) do not allow for revoking same certificate more than once.

The current implementation doesn't prevent that behavior. Every call to revoke a cert will add an entry to the CRL.
See lack of checking whether or not the cert has already been revoked in:
  * [puppetserver](https://github.com/puppetlabs/puppetserver/blob/master/src/clj/puppetlabs/puppetserver/certificate_authority.clj#L1275)
  * [jvm-ssl-utils](https://github.com/puppetlabs/jvm-ssl-utils/blob/master/src/java/com/puppetlabs/ssl_utils/SSLUtils.java#L346)
  * [puppet](https://github.com/puppetlabs/puppet/blob/master/lib/puppet/ssl/certificate_revocation_list.rb#L49)